File: src/main/java/com/sleepwalker/sleeplib/client/drawable/Drawable.java
package com.sleepwalker.sleeplib.client.drawable;

import com.sleepwalker.sleeplib.gg.essential.elementa.components.image.ImageProvider;
import com.sleepwalker.sleeplib.gg.essential.universal.UMatrixStack;

import javax.annotation.Nonnull;
import java.awt.*;

public interface Drawable extends ImageProvider {

        Drawable EMPTY = new Drawable() {
            @Override
            public void drawImage(@Nonnull UMatrixStack matrixStack, double x, double y, double width, double height, @Nonnull Color color) {}
            @Override
            public int getWidth() {
                return 0;
            }
            @Override
            public int getHeight() {
                return 0;
            }
        };

    @Override
    void drawImage(@Nonnull UMatrixStack matrixStack, double x, double y, double width, double height, @Nonnull Color color);

    @SuppressWarnings("deprecation")
    @Override
    default void drawImage(double x, double y, double width, double height, @Nonnull Color color) {
        drawImage(UMatrixStack.Compat.INSTANCE.get(), x, y, width, height, color);
    }

    @Override
    default void drawImageCompat(@Nonnull UMatrixStack matrixStack, double x, double y, double width, double height, @Nonnull Color color) {
        drawImage(matrixStack, x, y, width, height, color);
    }

    int getWidth();
    int getHeight();
}

File: src/main/kotlin/com/sleepwalker/sleeplib/elementa/drawable/Drawable.kt
package com.sleepwalker.sleeplib.elementa.drawable

import com.sleepwalker.sleeplib.gg.essential.elementa.components.image.ImageProvider
import com.sleepwalker.sleeplib.gg.essential.universal.UMatrixStack
import java.awt.Color

interface Drawable : ImageProvider {
    val width: Int
    val height: Int

    override fun drawImage(matrixStack: UMatrixStack, x: Double, y: Double, width: Double, height: Double, color: Color)

    companion object {
        val EMPTY = object : Drawable {
            override val width: Int
                get() = 0
            override val height: Int
                get() = 0

            override fun drawImage(matrixStack: UMatrixStack, x: Double, y: Double, width: Double, height: Double, color: Color) {}
        }
    }
}